### PR TITLE
avm1: Implement FileReferenceList

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**
## Description
Significantly progresses #308.
Supersedes #7637.

## Testing

Tests are included. 

## Checklist

- [x] I, a human, have self-reviewed this PR and fully understand the changes within.
- [x] I have [made or updated tests](https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#test-guidelines) where possible.
- [x] All of my commits are properly scoped, compile successfully, and pass all tests.
- [x] This PR does not make sense to split up into smaller PRs.
- [x] An LLM was involved in the authoring of this code.

### CODE REVIEW

**Comment by danielhjacobs** [comment]
Presumably supersedes https://github.com/ruffle-rs/ruffle/pull/7637?

**Comment by SuchAFuriousDeath** [comment]
> Presumably supersedes #7637?

It does indeed.

File: core/src/avm1/globals/utils.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+//! Shared utilities for AVM1 globals.
```
Comment (Lord-McSweeney) [inline core/src/avm1/globals/utils.rs]: I think the preferred method across the codebase is to keep this in `file_reference.rs` (assuming that was where it originally came from) and reference the method there in `file_reference_list.rs`

Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/utils.rs]: Sure, no problem

File: core/src/avm1/globals/file_reference.rs [hunk 2]
```diff
@@ -220,6 +224,83 @@ pub fn file_type<'gc>(
...
 }
 
+/// Helper error type so that `parse_file_filter_array` can use `?`.
+struct ReturnFalse;
+
+impl<'gc> From<Error<'gc>> for ReturnFalse {
+    fn from(_: Error<'gc>) -> Self {
+        Self
+    }
+}
```
Comment (moulins) [inline core/src/avm1/globals/file_reference.rs:241]: This is very suspicious; any exception raised by things like `coerce_to_string` will be swallowed, which is almost certainly wrong.

Instead, `parse_file_filter_array` should probably return something like `Result<None<Vec<_>>, Error<'gc>>`, with the `None` case representing the AS2 method returning `false`.

Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/file_reference.rs:241]: I think there's a test added for all the `false` cases and the error is really never returned, it's always `false`.

Comment (kjarosh) [reply core/src/avm1/globals/file_reference.rs:241]: Even if that's the case, instead of ignoring the error you can return it and ignore it in the caller (afer e.g. logging it).

Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/file_reference.rs:241]: Ok and what should be returned instead of the `return Err(ReturnFalse)`? There's no error there that's produced by an A method.

Comment (moulins) [reply core/src/avm1/globals/file_reference.rs:241]: > I think there's a test added for all the false cases and the error is really never returned, it's always false.

But I can't find any `throw` keyword in the new tests?

For example:
```as
FileReferenceList.browse({
  // your current code will swallow the exception, is this correct?
  extension: { toString: function() { throw "oops"; } },
});
```
Another edge-case that ideally should be tested is: what happens when the `extension`/etc members etc are dynamic properties or only present on the prototype?

> Ok and what should be returned instead of the return Err(ReturnFalse)? There's no error there that's produced by an A method.

A possibility would be returning `Ok(Some(file_filters))` when the filters are valid, and `Ok(None)` otherwise.

File: tests/tests/swfs/avm1/file_reference_list_browse_invalid_filters/test.as [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        trace("// empty description");
+        trace(list.browse([emptyDesc]));
+
+        var emptyExt = new Object();
+        emptyExt.description = "txt";
+        emptyExt.extension = "";
+        trace("// empty extension");
+        trace(list.browse([emptyExt]));
+
+        // The following four cases document a pre-existing oddity inherited
```
Comment (kjarosh) [inline tests/tests/swfs/avm1/file_reference_list_browse_invalid_filters/test.as]: What do you mean by pre-existing here? Every behavior we're testing is pre-existing

File: core/src/avm1/globals/file_reference_list.rs [hunk 4]
```diff
@@ -21,5 +30,60 @@ fn constructor<'gc>(
     _this: Object<'gc>,
     _args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
+    // `FileReferenceList` instances carry no native data; matching Flash, the
+    // `__proto__` slot stays writable from ActionScript.
```
Comment (kjarosh) [inline core/src/avm1/globals/file_reference_list.rs]: When is `__proto__` non-writable?

File: core/src/avm1/globals/file_reference.rs (hunk 2 above)
Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/file_reference.rs:241]: I will extend the test to include this.

File: tests/tests/swfs/avm1/file_reference_list_browse_invalid_filters/test.as [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        trace(list.browse([noExt]));
+
+        var emptyDesc = new Object();
+        emptyDesc.description = "";
+        emptyDesc.extension = "*.txt";
+        trace("// empty description");
+        trace(list.browse([emptyDesc]));
+
+        var emptyExt = new Object();
+        emptyExt.description = "txt";
```
Comment (kjarosh) [inline tests/tests/swfs/avm1/file_reference_list_browse_invalid_filters/test.as]: If you feel like it, you could also test cases where these properties have flags (e.g. non enumerable, version 10, etc.).

**Review by moulins — APPROVED** [review]

**Review by Lord-McSweeney — COMMENTED** [review]
Could you change `core::avm1::globals::file_reference::constructor` to use `FileReferenceObject::new` instead of manually constructing the `FileReferenceObject`?

**Comment by SuchAFuriousDeath** [comment]
> Could you change `core::avm1::globals::file_reference::constructor` to use `FileReferenceObject::new` instead of manually constructing the `FileReferenceObject`?

Done :saluting_face:

File: core/src/avm1/globals/file_reference.rs [hunk 6]
```diff
@@ -220,80 +224,85 @@ pub fn file_type<'gc>(
     Ok(Value::Undefined)
 }
 
-/// Helper error type so that `parse_file_filter_array` can use `?`.
-struct ReturnFalse;
```
Comment (kjarosh) [inline core/src/avm1/globals/file_reference.rs]: Should we squash? If not, why add and remove ReturnFalse?

Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/file_reference.rs]: I'll fix it up, sorry :(

File: core/src/avm1/globals/file_reference.rs [hunk 7]
```diff
@@ -220,6 +224,88 @@ pub fn file_type<'gc>(
...
+
+/// Parse the file filter argument used by `FileReference.browse` and
+/// `FileReferenceList.browse`.
+///
+/// Returns `Ok(Some(filters))` on success, or `Ok(None)` if `browse()` should
+/// bail out and return `false` without opening a dialog. Exceptions raised
+/// during property coercion are propagated.
+pub fn parse_file_filters<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    args: &[Value<'gc>],
```
Comment (Lord-McSweeney) [inline core/src/avm1/globals/file_reference.rs]: nit - It seems a bit odd to me to pass the entire args slice here- is there an advantage to this over this function simply taking a `Option<Value<'gc>>`, the first argument? Personally I would find it easier to understand the inputs of this function if it were taking only the first argument.

**Review by Lord-McSweeney — APPROVED** [review]

File: core/src/avm1/globals/file_reference.rs (hunk 7 above)
Comment (SuchAFuriousDeath) [reply core/src/avm1/globals/file_reference.rs]: Good point I guess, I will redo it.

**Comment by ROBERT-MCDOWELL** [comment]
@SuchAFuriousDeath the browse event still not working so I cannot select multiple files for upload. about download, it works but the progress event is not fired so the progressbar never reaches 100%
